share/wake/lib/core/map.wake

```diff
@@ -0,0 +1,548 @@
+#   ```
+export def mdelete (key: k) (map: Map k v): Map k v =
+    def deleteKey tree =
+        def Tree cmpPair root = tree
+        delete (map.getMapComparison _.getPairFirst _) key root
+        | Tree cmpPair
+    editMapData deleteKey map
+
+
+# Accumulate and combine every value in the map, starting from the "smallest" key.
+#
+# *Parameters:*
+#   * `fn`:
+#       The manner in which each value should be added to the accumulator.
+#   * `base`:
+#       The value used to initialize the accumulator.  If `map` is empty, this
+#       value is returned unchanged.
+#   * `map`:
+#       The key-value pairs which will be combined.
+#
+# *Examples:*
+#   ```
+#   mnew scmp | mfoldl (\_\a\v a + v) 0 = 0
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mfoldl (\_\a\v a + v) 0 = 3
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mfoldl (\k\a\v "{a} {k}={str v}") "k=v:" = "k=v: a=1 b=2"
+#   ```
+export def mfoldl (fn: k => a => v => a) (base: a) (map: Map k v): a =
+    def pairFn a (Pair k v) = fn k a v
+    tfoldl pairFn base map.getMapData
+
+# Accumulate and combine every value in the map, starting from the "largest" key.
+#
+# *Parameters:*
+#   * `fn`:
+#       The manner in which each value should be added to the accumulator.
+#   * `base`:
+#       The value used to initialize the accumulator.  If `map` is empty, this
+#       value is returned unchanged.
+#   * `map`:
+#       The key-value pairs which will be combined.
+#
+# *Examples:*
+#   ```
+#   mnew scmp | mfoldr (\_\v\a v + a) 0 = 0
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mfoldr (\_\v\a v + a) 0 = 3
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mfoldr (\k\v\a "{a} {k}={str v}") "k=v:" = "k=v: b=2 a=1"
+#   ```
+export def mfoldr (fn: k => v => a => a) (base: a) (map: Map k v): a =
+    def pairFn (Pair k v) a = fn k v a
+    tfoldr pairFn base map.getMapData
+
+# Transform and combine every value in the map in parallel.
+#
+# *Parameters:*
+#   * `combineFn`:
+#       The manner in which two values of the target type should be joined.
+#   * `base`:
+#       The value used to initialize the accumulator.  If `map` is empty, this
+#       value is returned unchanged.
+#   * `transformFn`:
+#       The function which should be applied to every key-value pair in the map.
+#       This might just be to prepare them to be combined, or it might be some
+#       more complex function which happens to have a more-easily-joined output.
+#   * `map`:
+#       The key-value pairs which will be processed.
+#
+# *Examples:*
+#   ```
+#   mnew scmp | mfoldmap (_+_) 0 (\_\v v) = 0
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mfoldmap (_+_) 0 (\_\v v) = 3
+#   listToMap scmp ("a" → 1, "b" → 2, Nil) | mfoldmap ("{_} {_}") "k=v:" ("{_}={str _}") = "k=v: a=1 b=2"
+#   ```
+export def mfoldmap (combineFn: a => a => a) (base: a) (transformFn: k => v => a) (map: Map k v): a =
+    def pairFn (Pair k v) = transformFn k v
+    tfoldmap combineFn base pairFn map.getMapData
```

Comment:
I wonder if a nicer interface is `combineFn: k => v => a => a`, I can see myself using `Pair` the `transformFn` a lot with this current interface.

Your interface does have a classic mapReduce style to it though which is nice.

Reply:
I guess foldl/foldr are just more generic than foldmap. You could implement this as follows right?

```
def fn k v a = combineFn (transformFn k v) a
mfoldr fn base map
```

So I guess this is fine, its just a decomposition some people might prefer.

Reply:
The main difference is in execution order; you could certainly reimplement the results and signature of foldmap, but it has a more branched behaviour where it executes the `combineFn` as soon as all neighbors have been completed, rather than adding one element at a time.